doc/notebooks/parallelization_tutorial.ipynb

```diff
@@ -0,0 +1,510 @@
+    "\n",
+    "# Start GRASS Session\n",
+    "session = gj.init(\"~/data/grassdata\", \"nc_basic_spm_grass7\", \"user1\")\n",
+    "\n",
+    "# Set computational region to the elevation raster.\n",
+    "gs.run_command(\"g.region\", raster=\"elevation\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e5f2a951-fff3-4ab5-b408-4e424e2834e4",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "## Using already parallelized tools\n",
+    "\n",
+    "There are many tools in GRASS GIS that are already parallelized, [see the list](https://grass.osgeo.org/grass-stable/manuals/keywords.html#parallel). Many tools in GRASS Addons are parallelized as well.\n",
+    "\n",
+    "Generally, there are two types of implementation.\n",
+    "Multithreading in C tools:\n",
+    "   * Threads have low overhead, so they can be spawned more efficiently.\n",
+    "   * Tools use OpenMP API. One of the advantages of OpenMP for software distribution is that code works (runs in serial) also without OpenMP library present on the system.\n",
+    "   * Memory is shared, so programmer needs to be cautious about race conditions (e.g., writing into the same variable).\n",
+    "   \n",
+    "Multiprocessing in Python tools:\n",
+    "   * There are multiple ways to implement it, typically tools use `subprocess` and `multiprocessing` package.\n",
+    "   * Python tools are often wrappers around GRASS tools implemented in C. For example, tool [r.sun.daily](https://grass.osgeo.org/grass-stable/manuals/addons/r.sun.daily.html) runs [r.sun](https://grass.osgeo.org/grass-stable/manuals/r.sun.html) for multiple days in parallel.\n",
+    "   \n",
+    "Parallelized tools have `nprocs` parameter to specify number of cores to use. For C tools using OpenMP, GRASS GIS needs to be compiled with OpenMP support to take advantage of it. Both implementations work well on a single machine, but can't be scaled to a distributed system.\n",
+    "   \n",
+    "Example of calling a parallelized module in Bash and comparing the time with using 1 and 4 cores: "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "2094699b-5693-4414-9a71-3bd360f0b5c9",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%bash\n",
+    "time r.neighbors --q input=elevation output=elevation_smooth size=25 method=average nprocs=1\n",
+    "time r.neighbors --q input=elevation output=elevation_smooth size=25 method=average nprocs=4"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "abb954e4-0d7e-4026-8e93-a2a3bee02214",
+   "metadata": {},
+   "source": [
+    "Scaling is not linear. The speedup (processing time with 1 core / processing time with N cores) decreases when adding cores. See [benchmarks for r.neighbors](https://grass.osgeo.org/grass-stable/manuals/r.neighbors.html#performance). This behavior is due to the serial parts of the code (see [Amdahl's law](https://en.wikipedia.org/wiki/Amdahl%27s_law)) and overhead. "
```

Comment:
Confusing, maybe: The speedup indeed decreases, but it does not decrease proportionally to cores added. ...or something like that. :-)